### PR TITLE
Add option to choose to which element append/prepend the banner

### PR DESCRIPTION
**Pull request by jinoraL**
Let me know if this is the PR you expected me to do

### CODE REVIEW

File: test/spec/smartbanner_spec.js [hunk 1]
```diff
@@ -52,12 +54,84 @@ describe('SmartBanner', function() {
...
+    </head>
+    <body>
+      <div id="js_target"></div>
+      <div class="ui-page ui-page-active" style="position:absolute; top:12px;"></div>
+      <div class="ui-page" style="position:absolute; top:13px;"></div>
+    </body>
+  </html>`;
+
+  const HTML_APPEND_TO_BODY =
+    `<!doctype html>
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's start on the same line of assignment.

File: test/spec/smartbanner_spec.js [hunk 2]
```diff
@@ -52,12 +54,84 @@ describe('SmartBanner', function() {
     </body>
   </html>`;
 
+  const HTML_PREPEND_TO_SPECIFIC_TARGET =
+    `<!doctype html>
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's start on the same line of assignment.

File: test/spec/smartbanner_spec.js [hunk 3]
```diff
@@ -8,7 +8,8 @@ import Bakery from '../../src/bakery.js';
 
 describe('SmartBanner', function() {
 
-  const HTML = `<!doctype html>
+  const HTML =
+    `<!doctype html>
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's start on the same line of assignment.

File: test/spec/smartbanner_spec.js [hunk 4]
```diff
@@ -30,7 +31,8 @@ describe('SmartBanner', function() {
     </body>
   </html>`;
 
-  const HTML_DISABLED_POSITIONING = `<!doctype html>
+  const HTML_DISABLED_POSITIONING =
+    `<!doctype html>
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's start on the same line of assignment.

File: src/smartbanner.js [hunk 5]
```diff
@@ -196,7 +205,13 @@ export default class SmartBanner {
...
+      document.querySelector(this.options.prependTarget).removeChild(banner);
+    } else if (this.options.appendTarget !== undefined) {
+      document.querySelector(this.options.appendTarget).removeChild(banner);
+    } else {
+      document.querySelector('body').removeChild(banner);
+    }
     Bakery.bake(this.hideTtl);
   }
-}
+}
```
Comment (ain) [inline src/smartbanner.js]: Please add EOL to the file, we have all files formatted by this convention.

File: src/smartbanner.js [hunk 6]
```diff
@@ -140,7 +140,8 @@ export default class SmartBanner {
   }
 
   get positioningDisabled() {
-    return this.options.disablePositioning === 'true';
+    return this.options.disablePositioning === 'true' || this.options.appendTarget !== undefined || this.options.prependTarget !==
+      undefined;
```
Comment (ain) [inline src/smartbanner.js]: Let's not break the return onto 2 lines.

File: test/spec/smartbanner_spec.js [hunk 7]
```diff
@@ -52,12 +54,84 @@ describe('SmartBanner', function() {
...
+      <meta name="smartbanner:append-target" content="body">
+    </head>
+    <body>
+      <div class="ui-page ui-page-active" style="position:absolute; top:12px;"></div>
+      <div class="ui-page" style="position:absolute; top:13px;"></div>
+    </body>
+  </html>`;
+
+  const HTML_APPEND_TO_SPECIFIC_TARGET =
+    `<!doctype html>
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's start on the same line of assignment.

File: test/spec/smartbanner_spec.js [hunk 8]
```diff
@@ -52,12 +54,84 @@ describe('SmartBanner', function() {
...
+  </html>`;
+
   const SCRIPTS = [
     'https://code.jquery.com/jquery-3.1.1.min.js',
     'https://code.jquery.com/mobile/1.4.5/jquery.mobile-1.4.5.min.js'
   ];
 
-  const IOS_BODY = `<div class="smartbanner smartbanner--ios js_smartbanner">
+  const IOS_BODY =
+    `<div class="smartbanner smartbanner--ios js_smartbanner">
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's start on the same line of assignment.

File: test/spec/smartbanner_spec.js [hunk 9]
```diff
@@ -70,7 +144,8 @@ describe('SmartBanner', function() {
       <a href="https://itunes.apple.com/us/genre/ios/id36?mt=8" target="_blank" class="smartbanner__button"><span class="smartbanner__button__label">View</span></a>
     </div>`;
 
-  const ANDROID_BODY = `<div class="smartbanner smartbanner--android js_smartbanner">
+  const ANDROID_BODY =
+    `<div class="smartbanner smartbanner--android js_smartbanner">
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's start on the same line of assignment.

File: test/spec/smartbanner_spec.js [hunk 10]
```diff
@@ -83,14 +158,22 @@ describe('SmartBanner', function() {
...
-  const USER_AGENT_IPHONE_CUSTOM_WEBAPP = 'Mozilla/5.0 (iPhone; CPU iPhone OS 9_3_2 like Mac OS X) AppleWebKit/601.1.46 (KHTML, like Gecko) Version/9.0 Mobile/13F69 Safari/601.1 My Example Webapp';
-  const USER_AGENT_IPAD = 'Mozilla/5.0 (iPad; CPU OS 9_3_2 like Mac OS X) AppleWebKit/601.1.46 (KHTML, like Gecko) Version/9.0 Mobile/13F69 Safari/601.1';
-  const USER_AGENT_IPOD = 'Mozilla/5.0 (iPod touch; CPU iPhone OS 8_4_1 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Version/8.0 Mobile/12H321 Safari/600.1.4';
-  const USER_AGENT_ANDROID = 'Mozilla/5.0 (Linux; Android 5.1; XT1039 Build/LPB23.13-17.6; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/50.0.2661.86 Mobile Safari/537.36 [FB_IAB/FB4A;FBAV/79.0.0.18.71;]';
-  const USER_AGENT_ANDROID_CUSTOM_WEBAPP = 'Mozilla/5.0 (Linux; Android 5.1; XT1039 Build/LPB23.13-17.6; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/50.0.2661.86 Mobile Safari/537.36 [FB_IAB/FB4A;FBAV/79.0.0.18.71;]  My Example Webapp';
-  const USER_AGENT_DESKTOP = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/601.7.7 (KHTML, like Gecko) Version/9.1.2 Safari/601.7.7';
+  const USER_AGENT_IPHONE_IOS8 =
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's start on the same line of assignment (incl. the below lines). Esp. not to generate a diff over auto-formatting.

File: test/spec/smartbanner_spec.js [hunk 11]
```diff
@@ -99,13 +182,16 @@ describe('SmartBanner', function() {
     context('without options', function() {
 
       before(function() {
-        global.document = jsdom.jsdom('<html></html>', { userAgent: USER_AGENT_IPHONE_IOS9 });
+        global.document = jsdom.jsdom('<html></html>', {
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's revert this, again, not to generate a diff in places where your feature is not relevant and should not touch down.

File: test/spec/smartbanner_spec.js [hunk 12]
```diff
@@ -99,13 +182,16 @@ describe('SmartBanner', function() {
...
+          userAgent: USER_AGENT_IPHONE_IOS9
+        });
         global.getComputedStyle = document.defaultView.getComputedStyle;
         smartbanner = new SmartBanner();
       });
 
       it('expected to throw error', function() {
-        expect(() => smartbanner.publish()).to.throw('No options detected. Please consult documentation.');
+        expect(() => smartbanner.publish()).to.throw(
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's revert this, again, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 13]
```diff
@@ -115,7 +201,9 @@ describe('SmartBanner', function() {
       context('when on iPhone', function() {
 
         beforeEach(function() {
-          global.window = jsdom.jsdom(HTML, {userAgent: USER_AGENT_IPHONE_IOS9}).defaultView;
+          global.window = jsdom.jsdom(HTML, {
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's revert this, again, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 14]
```diff
@@ -144,7 +232,9 @@ describe('SmartBanner', function() {
       context('when on iPad', function() {
 
         before(function() {
-          global.window = jsdom.jsdom(HTML, {userAgent: USER_AGENT_IPAD}).defaultView;
+          global.window = jsdom.jsdom(HTML, {
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's revert this, again, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 15]
```diff
@@ -161,7 +251,9 @@ describe('SmartBanner', function() {
       context('when on iPod', function() {
 
         before(function() {
-          global.window = jsdom.jsdom(HTML, {userAgent: USER_AGENT_IPOD}).defaultView;
+          global.window = jsdom.jsdom(HTML, {
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's revert this, again, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 16]
```diff
@@ -178,7 +270,9 @@ describe('SmartBanner', function() {
       context('when on Android', function() {
 
         before(function() {
-          global.window = jsdom.jsdom(HTML, { userAgent: USER_AGENT_ANDROID }).defaultView;
+          global.window = jsdom.jsdom(HTML, {
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's revert this, again, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 17]
```diff
@@ -195,7 +289,9 @@ describe('SmartBanner', function() {
       context('when on desktop', function() {
 
         before(function() {
-          global.window = jsdom.jsdom(HTML, { userAgent: USER_AGENT_DESKTOP }).defaultView;
+          global.window = jsdom.jsdom(HTML, {
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's revert this, again, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 18]
```diff
@@ -213,7 +309,9 @@ describe('SmartBanner', function() {
     context('when has been closed within current session', function() {
 
       before(function() {
-        global.window = jsdom.jsdom(HTML, { userAgent: USER_AGENT_ANDROID }).defaultView;
+        global.window = jsdom.jsdom(HTML, {
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's revert this, again, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 19]
```diff
@@ -231,7 +329,9 @@ describe('SmartBanner', function() {
     context('when has been closed, but is reopened in new session', function() {
 
       before(function() {
-        global.window = jsdom.jsdom(HTML, { userAgent: USER_AGENT_ANDROID }).defaultView;
+        global.window = jsdom.jsdom(HTML, {
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's revert this, again, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 20]
```diff
@@ -240,7 +340,9 @@ describe('SmartBanner', function() {
       });
 
       it('expected to add to body', function() {
-        global.window = jsdom.jsdom(HTML, { userAgent: USER_AGENT_ANDROID }).defaultView;
+        global.window = jsdom.jsdom(HTML, {
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's revert this, again, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 21]
```diff
@@ -251,7 +353,8 @@ describe('SmartBanner', function() {
     });
 
     context('when enabled-platform set to android, but opened on iOS', function() {
-      const HTML_WITH_PLATFROM_OPTION_ANDROID = `<!doctype html>
+      const HTML_WITH_PLATFROM_OPTION_ANDROID =
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's have assignment restored, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 22]
```diff
@@ -272,7 +375,9 @@ describe('SmartBanner', function() {
       </html>`;
 
       before(function() {
-        global.window = jsdom.jsdom(HTML_WITH_PLATFROM_OPTION_ANDROID, { userAgent: USER_AGENT_IPHONE_IOS9 }).defaultView;
+        global.window = jsdom.jsdom(HTML_WITH_PLATFROM_OPTION_ANDROID, {
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's have assignment restored, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 23]
```diff
@@ -285,8 +390,11 @@ describe('SmartBanner', function() {
 
     });
 
-    context('when enabled-platform set to android, but opened on iOS 9 which is included by include-user-agent-regex', function() {
-      const HTML_WITH_PLATFROM_OPTION_ANDROID_INCLUDE_IOS9 = `<!doctype html>
+    context(
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's have assignment restored, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 24]
```diff
@@ -285,8 +390,11 @@ describe('SmartBanner', function() {
...
+      'when enabled-platform set to android, but opened on iOS 9 which is included by include-user-agent-regex',
+      function() {
+        const HTML_WITH_PLATFROM_OPTION_ANDROID_INCLUDE_IOS9 =
+          `<!doctype html>
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's have assignment restored, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 25]
```diff
@@ -307,24 +415,29 @@ describe('SmartBanner', function() {
         </body>
       </html>`;
 
-      before(function() {
-        global.window = jsdom.jsdom(HTML_WITH_PLATFROM_OPTION_ANDROID_INCLUDE_IOS9, { userAgent: USER_AGENT_IPHONE_IOS9 }).defaultView;
-        global.document = window.document;
-        global.getComputedStyle = window.getComputedStyle;
-        smartbanner = new SmartBanner();
-      });
+        before(function() {
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Let's have assignment restored, not to generate a diff.

File: test/spec/smartbanner_spec.js [hunk 26]
```diff
@@ -669,4 +799,204 @@ describe('SmartBanner', function() {
...
+    });
+
+    it('expected banner to be removed on exit', function(done) {
+      smartbanner.exit();
+      let element = document.querySelector('.js_smartbanner');
+      expect(element).not.to.exist;
+      done();
+    });
+  });
+});
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: EOL missing.

**Comment by jinoraL** [comment]
Sorry, this was indeed due to the auto-formatting of my IDE.
I fixed this and updated the PR.

File: dist/smartbanner.js [hunk 27]
```diff
@@ -428,7 +428,15 @@ var SmartBanner = function () {
       }
 
       var bannerDiv = document.createElement('div');
-      document.querySelector('body').appendChild(bannerDiv);
+      if (this.options.prependTarget !== undefined) {
+        var parent = this.options.prependTarget;
+        document.querySelector(parent).insertBefore(bannerDiv, parent.firstChild);
```
Comment (ain) [inline dist/smartbanner.js]: I failed to get the prepending working properly. 

The test case:

__META__:
```html
  <meta name="smartbanner:prepend-target" content="#wrapper">
```

__HTML__:
```html
  <div id="wrapper">
    <div id="first"></div>
    <div id="last"></div>
  </div>
```

__Result__:
```html
<div id="wrapper">
    <div id="first"></div>
    <div id="last"></div>
  <div class="smartbanner smartbanner--ios js_smartbanner">
      <a href="javascript:void();" class="smartbanner__exit js_smartbanner__exit"></a>
      <div class="smartbanner__icon" style="background-image: url(http://a3.mzstatic.com/us/r30/Purple60/v4/c1/3b/b0/c13bb085-64c0-cc90-f97a-521b96963986/icon350x350.jpeg);"></div>
      <div class="smartbanner__info">
        <div>
          <div class="smartbanner__info__title">Smart Application</div>
          <div class="smartbanner__info__author">SmartBanner Contributors</div>
          <div class="smartbanner__info__price">FREE - On the App Store</div>
        </div>
      </div>
      <a href="https://itunes.apple.com/us/genre/ios/id36?mt=8" target="_blank" class="smartbanner__button"><span class="smartbanner__button__label">View</span></a>
    </div></div>
```

**Comment by ain** [comment]
Please also rebase to master, we have new features landing as part of `v1.8.0` in the meanwhile.

File: dist/smartbanner.js (hunk 27 above)
Comment (jinoraL) [reply dist/smartbanner.js]: I'm sorry, I must have had wrongly generated the minified file. I pushed the correct file and it should work properly.

**Comment by jinoraL** [comment]
@ain I will open a new PR as the repo from where I am doing my PR changed

**Comment by ain** [comment]
@jinoraL fair enough, but please follow my change requests here in your new PR so I can skip going through the same code.

**Comment by ain** [comment]
No activity, diverged. Closing.

**Pull request closed without merging**